### DISCUSSION TITLE
Some questions about rover's read_socket_data function

### DISCUSSION THREAD

**Discussion by spacewander** [apache/skywalking#10273] [Q&A] [answered]
First of all, thanks to @mrproliu for developing this project!

When reading the read_socket_data function, some questions come to my mind.

```
    if (size > MAX_PROTOCOL_SOCKET_READ_LENGTH) {
        size = MAX_PROTOCOL_SOCKET_READ_LENGTH;
    }
    asm volatile("%[size] &= 0x1f;\n" ::[size] "+r"(size) :);
    bpf_probe_read(&reader->buffer, size & MAX_PROTOCOL_SOCKET_READ_LENGTH, buf);
    reader->data_len = size & MAX_PROTOCOL_SOCKET_READ_LENGTH;
    return reader;
```

https://github.com/apache/skywalking-rover/blob/ab76f3e6cd4cbeb695780e315ca2ee28884398d8/bpf/profiling/network/args.h#L240-L246

The first question is, why do we use inline assembly to write `size &= 0x1f`?

The second question, since the check at the beginning guarantees `size <= MAX_PROTOCOL_SOCKET_READ_LENGTH`, do we need to add an extra `&= 0x1f` which seems to do the same thing?

The final question, as `size <= MAX_PROTOCOL_SOCKET_READ_LENGTH`, `size & MAX_PROTOCOL_SOCKET_READ_LENGTH` is always equal to `size`. So why don't we write `bpf_probe_read(&reader->buffer, size, buf);` directly?

**Comment by mrproliu** [comment, marked as answer]
I remember this relate to the eBPF verifier, you can get more detail on [this issue](https://github.com/iovisor/bcc/issues/1260). If we don't use the & operator to check the range of the size, the verifier will check failure. So, we need to check the size with the MAX_PROTOCOL_SOCKET_READ_LENGTH first, then use the & operator for verifier.  
For the final question, I think we could use the size directly. 😀

**Reply by spacewander** [reply]
Got it. But why do we use inline assembly to use the `&` instead of writing it in the C code?

**Reply by mrproliu** [reply]
It's still related to the eBPF verifier, but I cannot find the original issue. :(
We need to tell the eBPF we are using `&=` directly, otherwise the verifier still thinks we haven't checked the bound of the buffer length, then throw the `R2 unbounded memory access, use 'var &= const' or 'if (var < const)'` exception.

**Reply by spacewander** [reply]
Thanks for your clarification!
